### PR TITLE
Support Values in compiled runtime - part 2

### PR DESCRIPTION
**Pull request by henriknyman**
This is based on https://github.com/neo4j/neo4j/pull/11378 with additional commits from "Rework CompiledMaterializeValueMapper" onward.

- Introduce tracking of ValueType and AnyValueType as representation types in CodeGenType
- Let parameters stay as type AnyValue
- Let property values stay as type Value
- Let literal lists and maps be AnyValue
- Handle AnyValue/Value types internally in a bunch of places
- Use the same conversion to storable type as the interpreted runtime, which handles temporal and spatial types

NOTE: This is needed for the compiled runtime to properly handle the new temporal and spatial indexes.

### CODE REVIEW

File: community/values/src/test/java/org/neo4j/values/storable/UTF8StringValueTest.java [hunk 1]
```diff
@@ -116,10 +116,10 @@ public void shouldCompareTo()
                 int x = stringValue( string1 ).compareTo( utf8Value( string2.getBytes( UTF_8 ) ) );
                 int y = utf8Value( string1.getBytes( UTF_8 ) ).compareTo( stringValue( string2 ) );
                 int z = utf8Value( string1.getBytes( UTF_8 ) )
-                        .compareTo( utf8Value( string2.getBytes( UTF_8 ) ) );
+                         .compareTo( utf8Value( string2.getBytes( UTF_8 ) ) );
```
Comment (pontusmelke) [inline community/values/src/test/java/org/neo4j/values/storable/UTF8StringValueTest.java:119]: You probably need to update `TextValueSpecification` as well.

File: community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java [hunk 2]
```diff
@@ -45,141 +41,202 @@
...
-    public AnyValue mapRelationship( VirtualRelationshipValue value )
+    private static final class DryRunMaterializeValueMapper extends AbstractMaterializeValueMapper
     {
-        if ( value instanceof RelationshipProxyWrappingValue )
+        boolean needsConversion;
+
+        @Override
+        public AnyValue mapNode( VirtualNodeValue value )
         {
+            if ( value instanceof NodeValue )
```
Comment (pontusmelke) [inline community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java]: Maybe
```
{
    needsConversion = !value && value instanceof NodeReference;
    return value;
}
```

File: community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java [hunk 3]
```diff
@@ -45,141 +41,202 @@
...
-    }
 
-    @Override
-    public AnyValue mapPath( PathValue value )
-    {
-        if ( value instanceof PathWrappingPathValue )
+        @Override
+        public AnyValue mapRelationship( VirtualRelationshipValue value )
         {
+            if ( value instanceof RelationshipValue )
```
Comment (pontusmelke) [inline community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java]: Maybe
```
{
    needsConversion = !needsConversion && value instanceof RelationshipReference;
    return value;
}
```

File: community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledEquivalenceUtils.java [hunk 4]
```diff
@@ -262,6 +266,29 @@ public static int hashCode( Object[] array )
         }
     }
 
+    /**
+     * Calculate hash code of a AnyValue[] value
+     * @param array the value to compute hash code for
+     * @return the hash code of the given value
+     */
+    public static int hashCode( AnyValue[] array )
+    {
```
Comment (pontusmelke) [inline community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledEquivalenceUtils.java:276]: Can't we just use

```
Arrays.hashCode( array );
```

I don't like this trick.

**Review by pontusmelke — CHANGES_REQUESTED** [review]

File: community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledEquivalenceUtils.java (hunk 4 above)
Comment (henriknyman) [reply community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledEquivalenceUtils.java:276]: Only for this method or do you mean change all of the hashCode methods that use this?

File: community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java (hunk 2 above)
Comment (henriknyman) [reply community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java]: I think you mean `needsConversion || (value instanceof NodeReference)`

**Review by pontusmelke — APPROVED** [review]

**Pull request merged**